### PR TITLE
Basic CREATE USER implementation

### PR DESCRIPTION
**Pull request by Hydrocharged**
This adds the minimum for the `CREATE USER` statement to work, in that we have an in-memory representation of the grant tables, which may be accessed either directly (by the `mysql` database) or through the user statements (of which we only have `CREATE USER` for now).

### CODE REVIEW

File: sql/in_mem_table/inmem_table_data.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+package in_mem_table
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+// Eventually these structures should be replaced by memory tables, however at the time of writing this file, the
```
Comment (zachmu) [inline sql/in_mem_table/inmem_table_data.go:24]: If this is actually necessary (I don't believe it is, see the overall PR comments) then we should just bite the bullet and make memory tables fast / work with indexes, it's past time

**Review by zachmu — COMMENTED** [review]
This seems like a wrong turn, but I could be convinced otherwise.

There's no need to keep all physical grant / user rows in memory all the time. What you need is

1) A `Database` interface extension for `CheckPrivileges` (this is what your analyzer function will call into)
2) A `DatabaseProvider` interface extension for `GetUsers` / `GetGrants` etc., not sure exactly what's needed there

The table representation is only really relevant when querying the user / grant tables directly. And I don't know if we will ever care if that's fast, I'm totally fine constructing them from the primitives in 2) at runtime as needed. The thing that needs to be fast is auth during analysis, and for that you should provide custom hooks for integrators that they can make as fast as necessary.

This is the same pattern we have followed for all the other mysql system tables: we build them at query time from the information provided by integrators through the interfaces they implement. I don't see a good reason to deviate from that pattern for this.

Let me know if I have the wrong idea here.

File: sql/core.go [hunk 2]
```diff
@@ -284,6 +285,28 @@ type PartitionIter interface {
 	Next(*Context) (Partition, error)
 }
 
+// PartitionsToPartitionIter creates a PartitionIter that iterates over the given partitions.
+func PartitionsToPartitionIter(partitions ...Partition) PartitionIter {
```
Comment (zachmu) [inline sql/core.go:289]: Not an appropriate file to put this

File: sql/grant_tables/user.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+// RepresentedColumns implements the interface in_mem_table.InMemTableDataKey.
+func (u UserPrimaryKey) RepresentedColumns() []uint16 {
+	return userPkCols
+}
+
+var _ in_mem_table.InMemTableDataKey = UserPrimaryKey{}
+
+// createUserTable returns a new Grant Table representing the "user" table.
+func createUserTable() *grantTable {
```
Comment (zachmu) [inline sql/grant_tables/user.go:66]: Probably makes sense to define everything in here as a package level var (unless this function is going to eventually take arguments?)

**Review by zachmu — APPROVED** [review]
Seems generally fine, hard to give more concrete feedback until I see how it ties into an integrator

File: sql/grant_tables/user.go (hunk 3 above)
Comment (Hydrocharged) [reply sql/grant_tables/user.go:66]: Not planning on it taking arguments anytime soon. I don't quite want to pollute the namespace for every type and column that will be used (as I've got like 8 more tables to do), but I'll move a lot of this to an `init` function and store some of the results in package vars, which should be a happy middle ground.

**Pull request merged**